toolchain/check/eval.cpp

```diff
@@ -0,0 +1,193 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/eval.h"
+
+#include "toolchain/sem_ir/typed_insts.h"
+
+namespace Carbon::Check {
+
+// If the given instruction is constant, returns its constant value.
+static auto UnwrapIfConstant(Context& context, SemIR::InstId inst_id)
+    -> SemIR::InstId {
+  return context.constant_values().Get(inst_id);
+}
+
+// If the given instruction block contains only constants, returns a
+// corresponding block of those values.
+static auto UnwrapIfConstant(Context& context, SemIR::InstBlockId inst_block_id)
+    -> SemIR::InstBlockId {
+  auto insts = context.inst_blocks().Get(inst_block_id);
+  llvm::SmallVector<SemIR::InstId> const_insts;
+  for (auto inst_id : insts) {
+    auto const_inst_id = UnwrapIfConstant(context, inst_id);
+    if (!const_inst_id.is_valid()) {
+      return SemIR::InstBlockId::Invalid;
+    }
+
+    // Once we leave the small buffer, we know the first few elements are all
+    // constant, so it's likely that the entire block is constant. Resize to the
+    // target size given that we're going to allocate memory now anyway.
+    if (const_insts.size() == const_insts.capacity()) {
+      const_insts.reserve(insts.size());
+    }
+
+    const_insts.push_back(const_inst_id);
+  }
+  // TODO: If the new block is identical to the original block, return the
+  // original ID.
+  return context.inst_blocks().Add(const_insts);
+}
+
+// If the specified operands of the given instruction have constant values,
+// replaces the operands with their constant values and builds a corresponding
+// constant value. Otherwise returns `SemIR::InstId::Invalid`.
+template <typename InstT, typename... SpecificNodeId>
+static auto TryRebuildIfConstOperands(Context& context, SemIR::Inst inst,
```

Comment:
Should const/constant be consistently abbreviated on function names? e.g. UnwrapIfConstant vs TryRebuildIfConstOperands

Reply:
Switched to `Constant` for now, to match existing uses (eg, `GetConstantValue`). Perhaps we should reserve `const` for `const`-qualified types using the `const` keyword, rather than constant evaluation, but in any case, let's discuss that separately.

Reply:
FWIW, I think we already use `Addr` as an abbreviation of `Address` which does not specifically refer to the `addr` keyword. I'd be fine with `Const` in general on the same terms -- but fine for that to be separately handled due to existing cases.